Multiple Distinct-qualified Aggregation(Multi-DQA) MPP execution method

weinan003 opened the pull request:
This PR bring Multi-DQA MPP execution to master again after 9.6 merged but a new strategy introduced. A new aggregate, `AGG_TUP_SPLIT`, divides each input tuple to `n` output tuples (`n` is the number of DQA exprs).  Each output tuple only contains one DQA expr and all GROUP BY expr, so that we can remove multiple DQA in one path. After previous work, a normal two-stage aggregate without distinct attaches on top. 

For this method, it is a limitation that GROUP BY clause can not be the same as DQA expr. The two-stage aggregate will get the wrong result since  GROUP BY keeps its expr in every tuple. To deal with this corner case, I declare a new expr named ShadowExpr. It just simply copy its major expr's value and cheat executor.
```
SELECT count(distinct a), count(distinct b) FROM foo GROUP BY a;
```
this query will translate as
```
SELECT count(distinct a), count(distinct b) FROM foo GROUP BY a';
(a' is the Shadow Expr for a, and all (a')'s value == a value)
```

 
Co-authored-by: Adam Li <ali@pivotal.io>
Inspired by: Heikki Linnakangas <heikki.linnakangas@iki.fi>
## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

hlinnaka commented:
Nice work!

Looking at one of the example queries from the regression test:
```
explain (costs off) select count(distinct d), count(distinct dt) from dqa_t1;
                               QUERY PLAN                               
------------------------------------------------------------------------
 Finalize Aggregate
   ->  Gather Motion 3:1  (slice2; segments: 3)
         ->  Partial Aggregate
               ->  Streaming HashAggregate
                     Group Key: d, dt
                     ->  Redistribute Motion 3:3  (slice1; segments: 3)
                           Hash Key: d, dt
                           ->  TupleSplit
                                 Split by Col: d, dt
                                 ->  Seq Scan on dqa_t1
 Optimizer: Postgres query optimizer
(11 rows)
```
I was surprised to see a Streaming HashAggregate there. A streaming hash agg doesn't necessarily eliminate all the duplicates, it is just a "best-effort" deduplication step that deduplicates as much as it can, with the memory available. If the hash table fills up, it will discard old entries and produce duplicates.

Does the Partial Aggregate step do deduplication as well, or does it rely on the Streaming HashAggregate to do the deduplication? If it relies on the Streaming HashAggregate, it is wrong. If it does deduplication of its own, then the Streaming HashAggregate isn't doing anything useful, might as well let the Partial Aggregate handle it.

Having a Streaming HashAggregate *below* the Redistribute Motion would make sense, to eliminate as many duplicates as possible before sending them over the network.

hlinnaka commented:
The cost estimate for the TupleSplit node is not good:
```
postgres=#  explain analyze select count(distinct d), count(distinct dt) from dqa_t1;
                                                                     QUERY PLAN                                                                      
-----------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=4.20..4.21 rows=1 width=16) (actual time=2.126..2.126 rows=1 loops=1)
   ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=4.16..4.19 rows=1 width=16) (actual time=2.105..2.114 rows=3 loops=1)
         ->  Partial Aggregate  (cost=4.16..4.17 rows=1 width=16) (actual time=0.753..0.753 rows=1 loops=1)
               ->  Streaming HashAggregate  (cost=4.12..4.15 rows=1 width=8) (actual time=0.755..0.802 rows=24 loops=1)
                     Group Key: d, dt
                     Extra Text: (seg1)   Hash chain length 1.3 avg, 4 max, using 18 of 32 buckets; total 0 expansions.
                     ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=4.00..4.09 rows=1 width=8) (actual time=0.569..0.626 rows=86 loops=1)
                           Hash Key: d, dt
                           ->  TupleSplit  (cost=4.00..4.03 rows=1 width=8) (actual time=0.026..0.060 rows=80 loops=1)
                                 Split by Col: d, dt
                                 ->  Seq Scan on dqa_t1  (cost=0.00..4.00 rows=34 width=8) (actual time=0.023..0.036 rows=40 loops=1)
 Planning time: 1.232 ms
   (slice0)    Executor memory: 144K bytes.
   (slice1)    Executor memory: 59K bytes avg x 3 workers, 59K bytes max (seg0).
   (slice2)    Executor memory: 162K bytes avg x 3 workers, 168K bytes max (seg1).
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 16.441 ms
(18 rows)
```
The "rows" estimate of the TupleSplit node is 1, which is surely not right. I think the correct estimate would be "(# of rows from input) * (# of DQAs)".

weinan003 commented:
Thank you for helping point out these problems, I've updated commit and
fixed these.

Heikki Linnakangas <notifications@github.com> 于2019年12月11日周三 下午5:46写道：

> The cost estimate for the TupleSplit node is not good:
>
> postgres=#  explain analyze select count(distinct d), count(distinct dt) from dqa_t1;
>                                                                      QUERY PLAN
> -----------------------------------------------------------------------------------------------------------------------------------------------------
>  Finalize Aggregate  (cost=4.20..4.21 rows=1 width=16) (actual time=2.126..2.126 rows=1 loops=1)
>    ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=4.16..4.19 rows=1 width=16) (actual time=2.105..2.114 rows=3 loops=1)
>          ->  Partial Aggregate  (cost=4.16..4.17 rows=1 width=16) (actual time=0.753..0.753 rows=1 loops=1)
>                ->  Streaming HashAggregate  (cost=4.12..4.15 rows=1 width=8) (actual time=0.755..0.802 rows=24 loops=1)
>                      Group Key: d, dt
>                      Extra Text: (seg1)   Hash chain length 1.3 avg, 4 max, using 18 of 32 buckets; total 0 expansions.
>                      ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=4.00..4.09 rows=1 width=8) (actual time=0.569..0.626 rows=86 loops=1)
>                            Hash Key: d, dt
>                            ->  TupleSplit  (cost=4.00..4.03 rows=1 width=8) (actual time=0.026..0.060 rows=80 loops=1)
>                                  Split by Col: d, dt
>                                  ->  Seq Scan on dqa_t1  (cost=0.00..4.00 rows=34 width=8) (actual time=0.023..0.036 rows=40 loops=1)
>  Planning time: 1.232 ms
>    (slice0)    Executor memory: 144K bytes.
>    (slice1)    Executor memory: 59K bytes avg x 3 workers, 59K bytes max (seg0).
>    (slice2)    Executor memory: 162K bytes avg x 3 workers, 168K bytes max (seg1).
>  Memory used:  128000kB
>  Optimizer: Postgres query optimizer
>  Execution time: 16.441 ms
> (18 rows)
>
> The "rows" estimate of the TupleSplit node is 1, which is surely not
> right. I think the correct estimate would be "(# of rows from input) * (#
> of DQAs)".
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/9223?email_source=notifications&email_token=AC2QG4BAJGJZBMGERDZTDILQYCZHZA5CNFSM4JZLHFM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGSQERQ#issuecomment-564462150>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2QG4EVWYILVIGTFZAVEVDQYCZHZANCNFSM4JZLHFMQ>
> .
>

hlinnaka commented:
> Thank you for helping point out these problems, I've updated commit and fixed these.

Thanks, that was quick!

The new plan is:
```
postgres=#  explain analyze select count(distinct d), count(distinct dt) from dqa_t1;
                                                                  QUERY PLAN                                                                   
-----------------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=4.39..4.40 rows=1 width=16) (actual time=3.481..3.481 rows=1 loops=1)
   ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=4.35..4.38 rows=1 width=16) (actual time=3.359..3.455 rows=3 loops=1)
         ->  Partial Aggregate  (cost=4.35..4.36 rows=1 width=16) (actual time=2.373..2.373 rows=1 loops=1)
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=4.25..4.34 rows=1 width=8) (actual time=1.662..2.234 rows=39 loops=1)
                     Hash Key: d, dt
                     ->  Streaming HashAggregate  (cost=4.25..4.28 rows=1 width=8) (actual time=0.387..0.481 rows=34 loops=1)
                           Group Key: d, dt
                           Extra Text: (seg2)   Hash chain length 1.5 avg, 4 max, using 22 of 32 buckets; total 0 expansions.
                           ->  TupleSplit  (cost=4.00..4.25 rows=0 width=8) (actual time=0.052..0.180 rows=80 loops=1)
                                 Split by Col: d, dt
                                 ->  Seq Scan on dqa_t1  (cost=0.00..4.00 rows=34 width=8) (actual time=0.046..0.076 rows=40 loops=1)
 Planning time: 1.219 ms
   (slice0)    Executor memory: 144K bytes.
   (slice1)    Executor memory: 148K bytes avg x 3 workers, 148K bytes max (seg0).
   (slice2)    Executor memory: 104K bytes avg x 3 workers, 104K bytes max (seg0).  Work_mem: 33K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 5.391 ms
(18 rows)
```
The plan shape makes more sense now, with the Streaming Aggregate below the Redistribute Motion. But the "rows" estimate on the TupleSplit node and all the nodes above it are still wrong.

I feel that it would be better to have a completely new node type for the TupleSplit, instead of reusing the Agg node for it. It doesn't have much in common with Agg. It doesn't actually do any aggregation, its cost estimation is different. And the nodeAgg.c code is really complicated as it is, I'd rather not add any more new modes or states to it if we can avoid it. The TupleSplit node feels more similar to SplitUpdate, although it's not quite the same as that either.

hlinnaka commented:
This doesn't handle NULLs correctly. The TupleSplit node simply sets all the "other" columns to NULL, and those NULLs are passed to the aggregates. That works as long as the aggregate is strict, ie. it ignores NULLs, but not all aggregates are strict. I think all the built-in ones that also have a combine function are strict, but you can create one that isn't:
```
-- Like COUNT(col), but also counts NULLs
create or replace function countall_trans(c int, newval int) returns int as $$
  SELECT $1 + 1;
$$ language sql;
create aggregate countall(sfunc = countall_trans, basetype = int, stype = int, initcond = 0, combinefunc = int4pl);

-- Test table
create table nonullstab (a int, b int);
insert into nonullstab select 1, 1 from generate_series(1, 100);

-- This returns wrong result. countall(distinct a) should return 1. 
postgres=#  select countall(distinct a), count(distinct b) from nonullstab;
 countall | count 
----------+-------
        2 |     1
(1 row)
```
I think TupleSplit needs to produce another column, to indicate explicitly which DQA the row is for.

hlinnaka commented:
I haven't looked at it closely yet, but I've got a feeling that the ShadowEliminate stuff really shouldn't be needed. There's got to be a simpler way to do it. In the executor, after all the planning is done, if I understand correctly, the ShadowEliminate node is basically just doing a projection. Surely we don't need a new node type for that. The Agg node below it is perfectly capable of doing the projection by itself.

But I'd suggest that you fix the NULL issue first. I've got a hunch that once you fix that, a solution to the ShadowEliminate issue will jump out. Specifically, if the TupleSplit node adds a new column to the target list that indicates which DQA the row is for, you won't need ShadowExpr anymore.

weinan003 commented:
Yes, Heikki. You are right. I also realized this point when you mention
giving the id in TupleSplit id. I'm working on this patch, so far the
feature is done. But I need to run again icw to give me more confidence and
also some comments waiting for me to add. I think the new patch will finish
today or next Monday.

Heikki Linnakangas <notifications@github.com> 于2019年12月20日周五 上午3:04写道：

> I haven't looked at it closely yet, but I've got a feeling that the
> ShadowEliminate stuff really shouldn't be needed. There's got to be a
> simpler way to do it. In the executor, after all the planning is done, if I
> understand correctly, the ShadowEliminate node is basically just doing a
> projection. Surely we don't need a new node type for that. The Agg node
> below it is perfectly capable of doing the projection by itself.
>
> But I'd suggest that you fix the NULL issue first. I've got a hunch that
> once you fix that, a solution to the ShadowEliminate issue will jump out.
> Specifically, if the TupleSplit node adds a new column to the target list
> that indicates which DQA the row is for, you won't need ShadowExpr anymore.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/9223?email_source=notifications&email_token=AC2QG4C3EZDRGZDZNO5HHIDQZPATBA5CNFSM4JZLHFM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHKTGOQ#issuecomment-567620410>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2QG4DW2SYPMKXNODAXTXDQZPATBANCNFSM4JZLHFMQ>
> .
>

adam8157 commented on src/test/regress/sql/gp_dqa.sql:
```diff
@@ -63,6 +63,52 @@ explain (costs off) select count(distinct dqa_t1.dt) from dqa_t1, dqa_t2 where d
+explain (costs off) select corr(distinct d, i) from dqa_t1 group by c;
+
+-- multi args multidqa 
+select count(distinct c), corr(distinct d, i) from dqa_t1;
```
Does anyone know is this `UDA(distinct(col1, col2))` a SQL standard thing?

adam8157 replied:
Ah, this corr() is built-in, nice catch!

Ref: https://www.postgresql.org/docs/9.4/functions-aggregate.html#FUNCTIONS-AGGREGATE-STATISTICS-TABLE

xiong-gang commented on src/backend/executor/nodeTupleSplit.c:
```diff
@@ -0,0 +1,205 @@
+			return NULL;
+
+
+		slot_getallattrs(node->outerslot);
```
Wouldn't it be a little bit of waste to deform all the attributes? if the query only uses the first a few columns of a wide table...

adam8157 commented on src/backend/nodes/equalfuncs.c:
```diff
@@ -3737,6 +3737,9 @@ equal(const void *a, const void *b)
 			retval = _equalRoleSpec(a, b);
 			break;
+		case T_AggExprId:
+			retval = true;
```
Where does it perform equal() on this? Why it's always ignored?

xiong-gang commented:
The patch looks good to me in general.
I have a question: we only consider hash aggregate (two phase) to deduplicate the result of the tuple split. I think it might be good in many cases to consider group aggregate. The deduplication looks very like a grouping sets aggregation to me: `group by grouping sets ((expr id, DQA key1), (expr id, DQA key2))`. Is that correct?

weinan003 commented:
> The patch looks good to me in general.
> I have a question: we only consider hash aggregate (two phase) to deduplicate the result of the tuple split. I think it might be good in many cases to consider group aggregate. The deduplication looks very like a grouping sets aggregation to me: `group by grouping sets ((expr id, DQA key1), (expr id, DQA key2))`. Is that correct?

Thanks, Gang!  we can give another path to remove duplicate by Group Agg.

yydzero commented:
How is the performance impact after this PR? Thanks!

Eulerizeit commented:
On Sat, Feb 15, 2020 at 8:21 PM Yandong Yao <notifications@github.com>
wrote:

> How is the performance impact after this PR? Thanks!
>

We weren't actually doing performance testing on master until this and the
Grouping Sets PRs landed.

-- Rob


> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/9223?email_source=notifications&email_token=ACJYQZ6SBMENRSQB6MLA7G3RDCIKRA5CNFSM4JZLHFM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEL33C6Y#issuecomment-586658171>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACJYQZ7LCJO2WJWBUI3GU2TRDCIKRANCNFSM4JZLHFMQ>
> .
>

yydzero commented:
Do you mean https://github.com/greenplum-db/gpdb/pull/9219? Both PRs are merged, looking forward to seeing some performance numbers.
